### PR TITLE
Updated buildpack to use rustup.sh

### PR DESCRIPTION
**Pull request by binarycleric**
Previously this buildpack was trying to download and install rust manually which caused many stdlibs to be missing. Rust provides a script to automate a bunch of the install processes and switching to this greatly simplified the buildpack.

I tried it on a test app using iron and everything seemed to work correctly (https://github.com/binarycleric/resume-api). Forgive any bad code in that repo, I'm still learning Rust.

### CODE REVIEW

**Comment by emk** [comment]
Nice! I'll try to get some time to test this tomorrow morning, but if you
don't hear from me, feel free to ping me again.

**Comment by binarycleric** [comment]
Will do! Thanks.

**Comment by binarycleric** [comment]
Bump! :D

**Comment by emk** [comment]
Bump heard. :-) I have some free time again today to tackle this.

**Comment by binarycleric** [comment]
No rush! Thanks for your time.

File: bin/compile [hunk 1]
```diff
@@ -33,31 +33,29 @@ fi
...
 cd "$2"
 
+RUST_HOME=`pwd`/rust-$VERSION
+
 # Make sure we have the correct Rust binaries and set up PATH.
-if [ -d rust-cache-$VERSION ]; then
+if [ -d $RUST_HOME ]; then
   echo "-----> Using Rust version $VERSION"
 else
   echo "-----> Downloading Rust version $VERSION binaries from $URL"
```
Comment (RoPP) [inline bin/compile]: Maybe `echo "-----> Downloading rustup"` is more appropriate

Comment (binarycleric) [inline bin/compile]: Yes, a good point! I'll try to fix that up as soon as I get a chance. I just started a new job so I lot of my spoons have been going into that.

**Comment by binarycleric** [comment]
@emk Messaging has been updated. Sorry that took so long, I was investigating using https://github.com/rust-lang-nursery/rustup.rs as a possible replacement but I figure this could be a nice incremental improvement.

In the meantime I'll start working with rustup.rs more frequently and see if there is a way for us to further improve this buildpack.

cc @RoPP

**Comment by binarycleric** [comment]
@emk Just wanted to bump again. :) I'm also happy to help with maintaining this buildpack. I've been working with Rust a bit more and I do work at Heroku, just not on the languages team.

**Comment by emk** [comment]
OK, I'm sorry for the ridiculously slow response. :-( I kinda got buried under pull requests for a while, and I'm only now clearing them out.

Anyway, I'm going to do three things:

1. Go ahead and merge your PR, because it looks good and Rustup is awesome.
2. Test out the build pack.
3. Give you commit privileges on this repository since you submitted a nice patch and you work for Heroku. :-)

**Comment by emk** [comment]
OK, your PR is merged, I've done some testing and debugging to make sure the build pack is in good shape, I've updated the example app (and re-deployed it onto Cedar), and I've given you commit privileges in case you discover some future breakage.

I don't deploy many personal apps to Heroku these days, because I have containerized infrastructure on AWS these days that's a bit cheaper. But some of my clients use Heroku enthusiastically. So I'm happy to do some light, ongoing maintenance of this buildpack when I have time. But please feel free to commit changes yourself, or to merge pull requests that seem worthwhile. In the spirit of [I gave commit rights to someone I didn't know](http://jakewins.com/p/clickbait), it's time for me to share power a bit. 🙂

**Comment by binarycleric** [comment]
@emk Thanks for the response!  Don't worry about things taking too long, we all have lives outside of non-work Github after all. 😄 

I've got a few contacts in the rust community so I will try to stay synced up with them to ensure this remains current. Just so we're clear, this isn't part of my official duties at Heroku. I think Rust is really interesting and I'd like to work with it more often. What better way than to use the infrastructure I already know & love.

I'll read over that blog post when I have time. Feel free to reach out to me anytime if you have comments or concerns.

Thanks again and take care!

**Comment by binarycleric** [comment]
I also apologize for not updating the docs. 😢

**Comment by emk** [comment]
> Just so we're clear, this isn't part of my official duties at Heroku

Totally understood. :-) This isn't part of anybody's official duties, but if two people have commit rights, there's twice the chance that we'll fix it when it breaks.

Here's how I visualize this working: If you see something that should be fixed or improved, just go for it. Feel free to coordinate with me if I'm around, or to just go ahead and do whatever you think needs to be done. I'm happy to trust your good taste because your patches were fine and you use Heroku.

Rustup is working really great, and it implements #5, which was a long-standing wishlist feature for a lot of users. I think that there's one more "must have" feature, namely cached builds (#10) so we don't have to recompile everything from scratch. Cargo already has very good build caching; we just need to give it a directory.

**Comment by emk** [comment]
Oh, I've also given you commit rights on https://github.com/emk/heroku-rust-cargo-hello, which is the demo app for this buildpack.

And if you see that I'm "assigned" on an issue but not doing anything about it, feel free to steal if from me. I think that covers the basics, but if you have any further questions, please let me know.

**Comment by emk** [comment]
OK, I'm doing one more update pass over `bin/compile` to upgrade to the latest rustup and implement #10. I'll let you know when it's done.

**Comment by emk** [comment]
Your patch really allowed me make a lot of simplifications and implement some wishlist features. :-)

I've implemented #10 (build caching yay!), configured the buildpack to use `stable` Rust by default, and done another doc overhaul. At this point, I think this is probably the best heroku-buildpack-rust out there, and it should be at least somewhat future-proof.

See 	332615e for the most important changes to how Rustup works.

At this point, I think the next steps are:

1. If you have a chance and want to test my changes against one of your existing projects, you're welcome to let me know if I've broken it completely. 🤦
2. I announce to https://www.reddit.com/r/rust that this buildpack has been heavily modernized.

At this point, I don't plan to make any further improvements myself for a while. Please feel free to tweak things whenever you want.

**Comment by binarycleric** [comment]
Woohoo! Hopefully, this will help further Rust adoption on Heroku (a personal dream of mine). I'll look over your recent commit soon, but it looks good at a glance.

I'll test this against some of my example projects and let you know if I run into any issues. Hopefully, I will have time later this week.

**Pull request merged**